pkg/tbtc/signing_loop.go

```diff
@@ -0,0 +1,503 @@
+		attemptCounter:          0,
+		attemptStartBlock:       initialStartBlock,
+		attemptSeed:             attemptSeed,
+		doneCheck:               doneCheck,
+	}
+}
+
+// signingAttemptParams represents parameters of a signing attempt.
+type signingAttemptParams struct {
+	number                 uint
+	startBlock             uint64
+	timeoutBlock           uint64
+	excludedMembersIndexes []group.MemberIndex
+}
+
+// signingAttemptFn represents a function performing a signing attempt.
+type signingAttemptFn func(*signingAttemptParams) (*signing.Result, uint64, error)
+
+// signingRetryLoopResult represents the result of the signing retry loop.
+type signingRetryLoopResult struct {
+	// result if the outcome of the signing process.
+	result *signing.Result
+	// latestEndBlock is the block at which the slowest signer of the successful
+	// signing attempt completed signature computation. This block is also
+	// the common end block accepted by all other members of the signing group.
+	latestEndBlock uint64
+	// attemptTimeoutBlock is the block at which the successful attempt times
+	// out.
+	attemptTimeoutBlock uint64
+}
+
+// start begins the signing retry loop using the given signing attempt function.
+// The retry loop terminates when the signing result is produced or the ctx
+// parameter is done, whatever comes first.
+func (srl *signingRetryLoop) start(
+	ctx context.Context,
+	waitForBlockFn waitForBlockFn,
+	signingAttemptFn signingAttemptFn,
+) (*signingRetryLoopResult, error) {
+	for {
+		srl.attemptCounter++
+
+		// In order to start attempts >1 in the right place, we need to
+		// determine how many blocks were taken by previous attempts. We assume
+		// the worst case that each attempt failed at the end of the signing
+		// protocol.
+		//
+		// That said, we need to increment the previous attempt start
+		// block by the number of blocks equal to the protocol duration and
+		// by some additional delay blocks. We need a small cool down in
+		// order to mitigate all corner cases where the actual attempt duration
+		// was slightly longer than the expected duration determined by the
+		// signingAttemptMaximumProtocolBlocks constant.
+		//
+		// For example, the attempt may fail at the end of the protocol but the
+		// error is returned after some time and more blocks than expected are
+		// mined in the meantime.
+		if srl.attemptCounter > 1 {
+			srl.attemptStartBlock = srl.attemptStartBlock +
+				uint64(signingAttemptMaximumBlocks())
+		}
+
+		srl.logger.Infof(
+			"[member:%v] waiting for attempt [%v] start signal",
+			srl.signingGroupMemberIndex,
+			srl.attemptCounter,
+		)
+
+		announcementStartBlock := srl.attemptStartBlock + signingAttemptAnnouncementDelayBlocks
+		err := waitForBlockFn(ctx, announcementStartBlock)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed waiting for announcement start block [%v] "+
+					"for attempt [%v]: [%v]",
+				announcementStartBlock,
+				srl.attemptCounter,
+				err,
+			)
```

Comment:
Should we return with an error or log it and `continue`?

If we return an error, it will break the execution of `start` function in `signing.go` and will be logged as:
```
signingLogger.Errorf(
	"[member:%v] all retries for the signing failed; "+
		"giving up: [%v]",
	signer.signingGroupMemberIndex,
	err,
)
```
which is not true in this particular case.

Reply:
See 7405a712